apm-sniffer/apm-sdk-plugin/finagle-6.25.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/finagle/ClientTracingFilterInterceptor.java

```diff
@@ -0,0 +1,97 @@
+import org.apache.skywalking.apm.agent.core.context.trace.SpanLayer;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.EnhancedInstance;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.MethodInterceptResult;
+
+import java.lang.reflect.Method;
+
+import static org.apache.skywalking.apm.network.trace.component.ComponentsDefine.FINAGLE;
+import static org.apache.skywalking.apm.plugin.finagle.ContextHolderFactory.getLocalContextHolder;
+import static org.apache.skywalking.apm.plugin.finagle.ContextHolderFactory.getMarshalledContextHolder;
+import static org.apache.skywalking.apm.plugin.finagle.FinagleCtxs.SW_SPAN;
+
+public class ClientTracingFilterInterceptor extends AbstractInterceptor {
+
+    private static class CacheObjects {
+        private ContextHolder marshlledContextHolder;
+        private ContextHolder localContextHolder;
+
+        private CacheObjects(ContextHolder marshlledContextHolder, ContextHolder localContextHolder) {
+            this.marshlledContextHolder = marshlledContextHolder;
+            this.localContextHolder = localContextHolder;
+        }
+    }
+
+    @Override
+    public void beforeMethodImpl(EnhancedInstance enhancedInstance, Method method, Object[] objects, Class<?>[] classes,
+                                 MethodInterceptResult methodInterceptResult) throws Throwable {
+        ContextCarrier contextCarrier = new ContextCarrier();
+        AbstractSpan finagleSpan = ContextManager.createExitSpan("pending", contextCarrier, "");
+
+        finagleSpan.setComponent(FINAGLE);
+        SpanLayer.asRPCFramework(finagleSpan);
+
+        ContextHolder marshlledContextHolder = getMarshalledContextHolder();
+        marshlledContextHolder.let(SWContextCarrier$.MODULE$, SWContextCarrier.of(contextCarrier));
+
+        ContextHolder localContextHolder = getLocalContextHolder();
+        localContextHolder.let(SW_SPAN, finagleSpan);
+
+        enhancedInstance.setSkyWalkingDynamicField(new CacheObjects(marshlledContextHolder, localContextHolder));
+    }
+
+    @Override
+    public Object afterMethodImpl(EnhancedInstance enhancedInstance, Method method, Object[] objects, Class<?>[] classes, Object ret) throws Throwable {
+        CacheObjects cacheObjects = (CacheObjects) enhancedInstance.getSkyWalkingDynamicField();
+
+        final AbstractSpan finagleSpan = cacheObjects.localContextHolder.remove(SW_SPAN);
+        cacheObjects.marshlledContextHolder.remove(SWContextCarrier$.MODULE$);
+
+        finagleSpan.prepareForAsync();
+        ContextManager.stopSpan(finagleSpan);
+
+        ((Future<?>) ret).addEventListener(new FutureEventListener<Object>() {
+            @Override
+            public void onSuccess(Object value) {
+                finagleSpan.asyncFinish();
+            }
+
+            @Override
+            public void onFailure(Throwable cause) {
+                finagleSpan.errorOccurred();
+//                finagleSpan.log(cause);
```

Comment:
why comment this?

Reply:
```java
public ExitSpan log(Throwable t) {
        if (stackDepth == 1) {
            super.log(t);
        }
        return this;
    }
```
For ExitSpan, the log will only works when stackDepth=1.  I don't understand why, Does it ok to invoke log method here?

Reply:
> For ExitSpan, the log will only works when stackDepth=1. I don't understand why, Does it ok to invoke log method here?

Why your plugin works as depth > 1? And yes, this is designed in that way, meaning, if there is a client-side RPC works, the nested one wouldn't be captured.